src/databricks/labs/ucx/source_code/dependencies.py

```diff
@@ -105,13 +195,43 @@ def _load_source(self, object_info: ObjectInfo) -> str:
             return f.read().decode("utf-8")
 
 
+class PackageLoader(DependencyLoader):
```

Comment:
after a bit more torough analysis of this PR, i see that there has to be two (more) loader interfaces:

In fact, we need to have a `NotebookLoader` interface with `WorkspaceNotebookLoader` & `LocalNotebookLoader` implementations. 

Then we'll have to get the `FileLoader`, which is `sys.path` aware. `DependencyLoader` has to take them both as arguments. e.g. `load_dependency(self, dependency: Dependency, file_loader: FileLoader, notebook_loader: NotebookLoader)`, if we inject it via method.  Remember: workspace file is still a file, if we're executing code from a Databricks Workflow (or notebook). 

<img width="289" alt="image" src="https://github.com/databrickslabs/ucx/assets/259697/466b393c-5d20-4ee6-83ce-6aa71bcb911f">

Remember: we have to support running this for both local FS (when running `databricks labs ucx migrate-local-files`) and remote notebooks (e.g. when running `databricks labs ucx migrate-job-code --job-id ...`).

Full illustrative example:
<img width="597" alt="image" src="https://github.com/databrickslabs/ucx/assets/259697/8ec0f674-35ff-49f2-9fc3-9e6c450f1797">

Reply:
That's planned with #1360. Step-by-step...

Reply:
let's do it in this PR, otherwise people may rely on incorrect abstractions.

Reply:
Done